spec/OneClickInstall_spec.rb

```diff
@@ -44,20 +54,195 @@
+      it "Returns a list of the names of the packages currently selected for installation" do
+        packages_to_install = subject.GetRequiredPackages
+        expect(packages_to_install).to match_array(["vim"])
+      end
+    end
+  end
+
+  describe "#GetRequiredPatterns" do
+    context "get list of patterns to install" do
+      it "Returns a list of the names of the patterns currently selected for installation" do
+        patterns_to_install = subject.GetRequiredPatterns
+        expect(patterns_to_install).to match_array(["x11"])
+      end
+    end
+  end
+
+  describe "#HaveAnyRecommended" do
+    context "return if there are recommended repositories or software" do
+      it "Returns boolean, depending on the existence of any recommended repositories or software" do
+        any_recommended = subject.HaveAnyRecommended
+        expect(any_recommended).to be true
+      end
+    end
+  end  
+
+  describe "#makeXMLFriendly" do
+    context "convert map to list of maps" do
+      it "converts map structure (eg. @repositories) to a list of maps with a key element" do
+        flattened = subject.makeXMLFriendly(software)
+        expect(flattened).to match_array([
+          {"recommended"=>"true", "action"=>"remove", "type"=>"pattern", "key"=>"sap_server"},
+          {"recommended"=>"false", "action"=>"install", "type"=>"package", "key"=>"tftp"},
+          {"recommended"=>"true", "action"=>"install", "type"=>"package", "key"=>"vim"},
+          {"recommended"=>"true", "action"=>"install", "type"=>"pattern", "key"=>"x11"}])
+      end
+    end
+  end
+
+  describe "#fromXMLFriendly" do
+    context "convert list of maps to map" do
+      it "converts back from makeXMLFriendly to original structure" do
+        flattened = [
+          {"recommended"=>"true", "action"=>"remove", "type"=>"pattern", "key"=>"sap_server"},
+          {"recommended"=>"false", "action"=>"install", "type"=>"package", "key"=>"tftp"},
+          {"recommended"=>"true", "action"=>"install", "type"=>"package", "key"=>"vim"},
+          {"recommended"=>"true", "action"=>"install", "type"=>"pattern", "key"=>"x11"}]
+        unflattened = subject.fromXMLFriendly(flattened)
+        expect(unflattened).to match_array(software)
+      end
+    end
+  end
+
+  describe "#ToXML" do
+    context "serialize OneClickInstall data structure" do
+      it "serializes to XML file, according to OneClickInstall data structure" do
+        filename = File.join(DATA_PATH, "serialized_result.xml")
+        subject.ToXML(filename)
+        expected = File.join(DATA_PATH, "serialized_expected.xml") 
+        expect(IO.read(filename)).to eql(IO.read(expected))
+        File.delete(filename)
+      end
+    end
+  end
+
+  describe "#FromXML" do
+    context "deserialize OneClickInstall data structure" do
+      it "deserializes according to OneClickInstall data structure to XML" do
+        filename = File.join(DATA_PATH, "to_deserialize.xml")
```

Comment:
wondering why you use different file here, just curious.

Reply:
Because I wanted the file "to_deserialize" to contain different information, so that we know for sure that the finction FromXML indeed causes changes. I hope that makes sense.

Reply:
I assumed you wonder why I didn't use the file "serialized_expected" that is already there. That file contains data that derive from the existing info in the instance variables, as set by "let". So, in case FromXML function does nothing at all, the content of the variables will be left as set by let functions and the example will succeed, even though the function doesn't work.

Reply:
ahm I missed that, thanks so then it should be better to initialize to nil before running the method that to maintain extra file, but just my point of view what is cheaper :)

Reply:
Would do. It would be cheaper in files but then, more commands. Also, the content of the software and repositories variables would be bigger.

Reply:
either way is fine, thanks for checking.